framework/test/unit/state_machine/state_machine.spec.ts

```diff
@@ -206,28 +227,52 @@ describe('state_machine', () => {
 			expect(events[0].toObject().topics[0]).toEqual(transaction.id);
 		});
 
+		it('should rollback state if beforeCommandExecute fails', async () => {
+			for (const e of testEvents) {
+				eventQueue.add(e.moduleID, e.typeID, e.data, e.topics);
+			}
+
+			const snapshotID = eventQueue.createSnapshot();
+
+			const ctx = new TransactionContext({
+				eventQueue,
+				logger,
+				stateStore,
+				header,
+				assets,
+				networkIdentifier,
+				transaction,
+				currentValidators: [],
+				impliesMaxPrevote: true,
+				maxHeightCertified: 0,
+				certificateThreshold: BigInt(0),
+			});
+			await stateMachine.executeTransaction(ctx);
+
+			eventQueue.add(
+				utils.intToBuffer(3, 4),
+				Buffer.from([0, 0, 0, 1]),
+				utils.getRandomBytes(100),
+				[utils.getRandomBytes(32)],
+			);
+
+			systemMod.beforeCommandExecute.mockImplementation(() => {
+				throw new Error('beforeCommandExecute failed');
+			});
+			mod.beforeCommandExecute.mockImplementation(() => {
+				throw new Error('beforeCommandExecute failed');
+			});
```

Comment:
If I understand the test scenario correctly
I think these functions are called within the `await stateMachine.executeTransaction(ctx);` so it should be mocked before calling them?
Also, instead of failing in `before` hook, i think we can add an event here, and fail on `after` then check if the end result is the same as original state?